File: src/main/scala/laoflch/debezium/connector/informix/InformixCDCEngine.scala
package laoflch.debezium.connector.informix

import com.informix.jdbcx.IfxDataSource
import com.informix.stream.api.IfmxStreamRecord
import com.informix.stream.cdc.IfxCDCEngine
import com.informix.stream.impl.IfxStreamException
import io.debezium.relational.TableId
import laoflch.debezium.connector.informix.InformixCDCEngine.{CDCTabeEntry, watchAllTableAndCols}

import java.sql.SQLException
import scala.jdk.CollectionConverters
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable

object InformixCDCEngine {

  private val URL_PATTERN: String = "jdbc:informix-sqli://%s:%s/syscdcv1:user=%s;password=%s;"
  // private val URL_PATTERN: String = "jdbc:informix-sqli://%s:%s/syscdcv1:user=%s;password=%s;PROTOCOLTRACE=2;PROTOCOLTRACEFILE=/tmp/dbz_proto_trace.out;TRACE=3;TRACEFILE=/tmp/dbz_trace.out"

  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[InformixCDCEngine])

  def genURLStr(host: String, port: String, dataBase: String, user: String, password: String): String = {
    URL_PATTERN.format(host, port.toString, user, password)
  }

  /*def initDataSource(url :String) : IfxDataSource = {
     new IfxDataSource(url)
  }

  def initEngine(dataSource:IfxDataSource): Unit={

    val builder=new IfxCDCEngine.Builder(dataSource)
    builder.build.init

  }*/

  def watchAllTableAndCols(watchTables: Map[String, CDCTabeEntry], builder: IfxCDCEngine.Builder): Unit = {
    watchTables.foreach(tuple => builder.watchTable(tuple._1, tuple._2.tableCols: _*))
    //foreach  says argument is tuple -> unit, so We can easily do below
    //https://stackoverflow.com/questions/8610776/scala-map-foreach for more info
  }

  def build(host: String, port: String, user: String, dataBase: String, password: String): InformixCDCEngine = {
    new InformixCDCEngine(host, port, user, dataBase, password)
  }

  case class CDCTabeEntry(tableId: TableId, tableCols: Seq[String])
}

class InformixCDCEngine(host: String, port: String, user: String, dataBase: String, password: String) {

  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[InformixCDCEngine])

  var cdcEngine: IfxCDCEngine = null
  var lsn: Long = 0x00
  var timeOut: Int = 0x05
  var hasInit: Boolean = false
  var labelId_tableId_map: Map[Int, TableId] = null

  def init(schema: InformixDatabaseSchema): Unit = {

    val url = InformixCDCEngine.genURLStr(host, port, dataBase, user, password)
    val url_masked = InformixCDCEngine.genURLStr(host, port, dataBase, user, "****")
    this.cdcEngine = this.buildCDCEngine(url, this.lsn, timeOut, schema)

    LOGGER.info("Connecting to Informix URL: {}", url_masked)

    try {
      this.cdcEngine.init()
      hasInit = true
    } catch {
      case e: SQLException =>
        e.printStackTrace()
      case e: IfxStreamException =>
        e.printStackTrace()
    }
  }

  def buildCDCEngine(url: String, lsn: Long, timeOut: Int,
                     schema: InformixDatabaseSchema): IfxCDCEngine = {

    val builder = new IfxCDCEngine.Builder(new IfxDataSource(url))

    LOGGER.info("CDCEngine set LSN = {}", lsn)

    // TODO: Make an parameter 'buffer size' for better performance. Default value is 10240
    builder.buffer(819200)

    /*
     * Watch all tables and build a Map for looking up "label_id"
     */
    val tableIds = CollectionConverters.SetHasAsScala(schema.tableIds()).asScala.toList
    tableIds.foreach(tid => {
      val cols = CollectionConverters.ListHasAsScala(schema.tableFor(tid).columns()).asScala.map(col => col.name()).toList
      val tname = tid.catalog() + ":" + tid.schema() + "." + tid.table()
      builder.watchTable(tname, cols: _*)
    })

    if (lsn > 0) {
      builder.sequenceId(lsn)
    }
    builder.timeout(timeOut)

    for (tbl <- CollectionConverters.ListHasAsScala(builder.getWatchedTables).asScala.toList) {
      LOGGER.info("Watched Tables :: {} ==> {}", tbl, tbl.getColumns)
    }
    this.labelId_tableId_map = tableIds.zipWithIndex.map { case (tid, idx) =>
      // TODO: label_id start from 1?
      (idx + 1, tid)
    }.toMap

    builder.build()
  }

  def record(func: (IfmxStreamRecord) => Boolean): Unit = {
    func(cdcEngine.getRecord)
  }

  def stream(func: (IfmxStreamRecord) => Boolean): Unit = {
    while (func(cdcEngine.getRecord)) {
      //Thread.sleep(1000)
    }
  }

  def setStartLsn(startLsn: Long): Long = {
    lsn = startLsn
    lsn
  }

  def convertLabel2TableId(): Map[Int, TableId] = {

    // LOGGER.info("converLabel2TableId(), tables={}", this.tables)
    LOGGER.info("convertLabel2TableId, LabelId => TableId :: {}", this.labelId_tableId_map)

    /*
    this.tables.map[(Int, TableId)](x => {
      val id = x.getDatabaseName + ":" + x.getNamespace + ":" + x.getTableName
      x.getLabel -> this.tableColsMap(id).tableId
    }).toMap

    */

    this.labelId_tableId_map
  }

  // def setTableColsMap(tableColsMap: Map[String, CDCTabeEntry]): Unit = this.tableColsMap = tableColsMap

}

File: src/main/java/laoflch/debezium/connector/informix/InformixCDCEngine.java
package laoflch.debezium.connector.informix;

import com.informix.jdbcx.IfxDataSource;
import com.informix.stream.api.IfmxStreamRecord;
import com.informix.stream.cdc.IfxCDCEngine;
import com.informix.stream.impl.IfxStreamException;
import io.debezium.config.Configuration;
import io.debezium.jdbc.JdbcConfiguration;
import io.debezium.relational.TableId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class InformixCDCEngine {

    private static Logger LOGGER = LoggerFactory.getLogger(InformixCDCEngine.class);

    private static String URL_PATTERN = "jdbc:informix-sqli://%s:%s/syscdcv1:user=%s;password=%s;";

    private String host;
    private String port;
    private String user;
    private String password;

    long lsn;
    int timeOut;
    boolean hasInit;

    private IfxCDCEngine cdcEngine;

    private Map<Integer, TableId> labelId_tableId_map;

    public InformixCDCEngine(Configuration config) {
        cdcEngine = null;
        lsn = 0;
        timeOut = 5;
        hasInit = false;

        host = config.getString(JdbcConfiguration.HOSTNAME);
        port = config.getString(JdbcConfiguration.PORT);
        user = config.getString(JdbcConfiguration.USER);
        password = config.getString(JdbcConfiguration.PASSWORD);
        // new InformixCDCEngineJava(host, port, user, password);

        // TODO: try HPPC or FastUtils's Integer Map?
        labelId_tableId_map = new Hashtable<>();
    }

    public void init(InformixDatabaseSchema schema) {
        try {
            String url = InformixCDCEngine.genURLStr(host, port, user, password);
            this.cdcEngine = this.buildCDCEngine(url, this.lsn, this.timeOut, schema);

            LOGGER.info("Connecting to Informix URL: {}", url);

            this.cdcEngine.init();
            this.hasInit = true;
        } catch (SQLException ex) {
            LOGGER.error("Caught SQLException: {}", ex);
        } catch (IfxStreamException ex) {
            LOGGER.error("Caught IfxStreamException: {}", ex);
        }
    }

    public IfxCDCEngine buildCDCEngine(String url, Long lsn, int timeOut, InformixDatabaseSchema schema) throws SQLException {
        IfxDataSource ds = new IfxDataSource(url);
        IfxCDCEngine.Builder builder = new IfxCDCEngine.Builder(ds);

        // TODO: Make an parameter 'buffer size' for better performance. Default value is 10240
        builder.buffer(819200);

        schema.tableIds().stream().forEach((TableId tid) -> {
            List<String> cols = schema.tableFor(tid).columns().stream()
                    .map(col -> col.name())
                    .collect(Collectors.toList());

            String tname = tid.catalog() + ":" + tid.schema() + "." + tid.table();
            builder.watchTable(tname, cols.toArray(new String[0]));
        });

        if (lsn > 0) {
            builder.sequenceId(lsn);
        }
        builder.timeout(timeOut);

        /*
         * Build Map of Label_id to TableId.
         */
        for (IfxCDCEngine.IfmxWatchedTable tbl: builder.getWatchedTables()) {
            TableId tid = new TableId(tbl.getDatabaseName(), tbl.getNamespace(), tbl.getTableName());
            labelId_tableId_map.put(tbl.getLabel(), tid);
            LOGGER.info("Added WatchedTable : label={} -> tableId={}", tbl.getLabel(), tid);
        }

        return builder.build();
    }

    public Long setStartLsn(Long fromLsn) {
        this.lsn = fromLsn;
        return fromLsn;
    }

    public Map<Integer, TableId> convertLabel2TableId() {
        return this.labelId_tableId_map;
    }

    public void stream(StreamHandler streamHandler) throws InterruptedException, SQLException, IfxStreamException {
        while (streamHandler.accept(cdcEngine.getRecord())) {

        }
    }

    public IfxCDCEngine getCdcEngine() {
        return cdcEngine;
    }

    public static String genURLStr(String host, String port, String user, String password) {
        return String.format(URL_PATTERN, host, port, user, password);
    }

    public static InformixCDCEngine build(Configuration config) {
        return new InformixCDCEngine(config);
    }

    public interface StreamHandler {

        boolean accept(IfmxStreamRecord record) throws SQLException, IfxStreamException, InterruptedException;

    }
}
